feat: implement multiple configuration files support via STARSHIP_CONFIG environment variable

Daniel-Escamilla opened the pull request:
## Description
Adds support for multiple configuration files via the STARSHIP_CONFIG environment variable. Files are merged in order, using `:` as a separator on Unix and `;` on Windows. Later files override earlier ones, maintaining backward compatibility.

## Motivation and Context  
Improves flexibility by allowing configurations to be split into reusable files (e.g., shared base plus local overrides). Helps organize setups across different environments.

Closes #6555

## Screenshots (if appropriate):
Unix example:  
`export STARSHIP_CONFIG="~/.config/starship/base.toml:~/.config/starship/custom.toml"`

Windows example:  
`$env:STARSHIP_CONFIG="C:\Users\User\.config\starship\base.toml;C:\Users\User\.config\starship\custom.toml"`

## How Has This Been Tested?

### Automated Tests:
```
Multiple files with platform separators (; Windows, : Unix)
Partial file existence scenarios (first-only, second-only, none-exist)
TOML merging precedence (later files override earlier)
Cross-platform separator validation and error handling
Fallback to default config when files missing
```

### Manual Testing:

#### Windows

``` powershell
$env:STARSHIP_CONFIG = 'config1.toml;config2.toml;config3.toml'
Invoke-Expression (&starship init powershell)
```

#### Unix

``` sh
export STARSHIP_CONFIG='config1.toml:config2.toml:config3.toml'
eval "$(starship init zsh)"
```

### UTF-8 Fix for Windows CMD (Commit fc1e3dca)
```
Custom modules with international characters display correctly
UTF-8 codepage (65001) properly set before command execution
```
### Backward compatibility with ASCII commands maintained
```
Memory Management Refactor (Commit f66737f4)
Compile-time safety with cargo check and cargo clippy
No performance regression in benchmarks
All lifetime-annotated modules function correctly across platforms
```
- [x] I have tested using **Linux**
- [x] I have tested using **Windows**
- [ ] I have tested using MacOS

### Checklist:
- [x] I have updated the tests accordingly.
- [x] I have updated the documentation accordingly.

Daniel-Escamilla committed:
feat(install.sh): correctly handle negative responses in installation directory prompt

Daniel-Escamilla committed:
fix(install.sh): deduplicate interrupt trap and fix spacing

Daniel-Escamilla committed:
Merge branch 'starship:master' into master

Daniel-Escamilla committed:
feat(install.sh): improve UX with example path in custom install prompt.

Daniel-Escamilla committed:
feat(install/install.sh): Small teaks in UX

Daniel-Escamilla committed:
feat: implement multiple configuration files support with TOML merging

Add support for multiple config files via STARSHIP_CONFIG using platform separators
(';' for Windows, ':' for Unix). Implements TOML merging with proper precedence,
comprehensive test coverage, and shell script integration for automatic config setup.

Closes #6555

Daniel-Escamilla committed:
fix: add UTF-8 codepage setting for Windows CMD commands in custom modules

Set UTF-8 codepage (65001) for Windows CMD before executing custom module commands
to fix character encoding issues with international characters.

Daniel-Escamilla committed:
refactor: add lifetime annotations for better memory management

Add explicit lifetime parameters to function signatures across multiple modules
for improved code clarity and memory safety. Minor code refinements and
shell script integration improvements.

Daniel-Escamilla committed:
refactor: simplify path separator detection using std::env::split_paths

Daniel-Escamilla commented:
Hi @davidkna 
I’ve updated the code to use std::env::split_paths like you mentioned.
Right now my version creates a merged config file beforehand, while #1887 does the merging at runtime. Do you think I should switch mine to merge at runtime too?
Just want to be sure before moving forward. Thanks!

cappyzawa commented on src/context.rs:
```diff
@@ -462,20 +466,229 @@ impl Default for Context<'_> {
+                    .insert(key, cleaned);
+            }
         }
     }
```
Great feature! 🎉 This addresses a real need for splitting configurations across multiple files. However, I'm concerned about the custom TOML parser in `src/context.rs:507-531`. It fails with complex TOML structures like arrays:

```toml
# config1.toml
[[battery.display]]
threshold = 10
style = "bold red"

# config2.toml  
[[battery.display]]
threshold = 25
style = "bold yellow"
```

The current line-by-line parser can't handle `[[battery.display]]` syntax properly. Since Starship already uses `toml = "0.9.5"`, consider using `toml::from_str()` for parsing instead of the custom approach. This would handle arrays, nested tables, and multiline strings properly.

cappyzawa reviewed (changes requested):
Great feature! 🎉 This addresses a real need for splitting configurations across multiple files. However, I'm concerned about the custom TOML parser in `src/context.rs:507-531`. It fails with complex TOML structures like arrays:

```toml
# config1.toml
[[battery.display]]
threshold = 10
style = "bold red"

# config2.toml  
[[battery.display]]
threshold = 25
style = "bold yellow"
```

The current line-by-line parser can't handle `[[battery.display]]` syntax properly. Since Starship already uses `toml = "0.9.5"`, consider using `toml::from_str()` for parsing instead of the custom approach. This would handle arrays, nested tables, and multiline strings properly.

cappyzawa commented:
Here's a suggested implementation using the existing `toml` crate:

```rust
fn merge_toml_files_proper(paths: &[PathBuf]) -> Option<toml::Table> {
    let mut merged = toml::Table::new();
    for path in paths {
        let content = fs::read_to_string(path).ok()?;
        let table: toml::Table = toml::from_str(&content).ok()?;
        merge_tables(&mut merged, table);
    }
    Some(merged)
}

fn merge_tables(base: &mut toml::Table, overlay: toml::Table) {
    for (key, value) in overlay {
        match (base.get_mut(&key), &value) {
            // Both are tables → merge recursively  
            (Some(toml::Value::Table(base_table)), toml::Value::Table(overlay_table)) => {
                merge_tables(base_table, overlay_table.clone());
            }
            // Otherwise → overwrite
            _ => { base.insert(key, value); }
        }
    }
}
```

This approach properly handles complex TOML structures like `[[battery.display]]` arrays and nested sections while preserving the existing functionality.

Daniel-Escamilla committed:
refactor: move multi-file config merging to runtime and update bug report output

- Shifts TOML merging logic for multiple STARSHIP_CONFIG files to runtime, improving flexibility and maintainability.
- Updates config loading, editing, and bug report generation to reflect the actual merged configuration in use.
- Adjusts and extends related tests for the new runtime merging approach.

Daniel-Escamilla committed:
refactor: simplify shell initialization by cleaning up print_script logic

- Removes legacy config merging logic from shell init scripts.
- Ensures shell initialization is now fully compatible with runtime config merging.

Daniel-Escamilla committed:
Merge branch 'starship:master' into master

Daniel-Escamilla committed:
Merge branch 'master' into feature/config-flag-6555

Daniel-Escamilla committed:
test(config, bug_report): cover error handling and config env cases in tests

Daniel-Escamilla committed:
Merge branch 'starship:master' into feature/config-flag-6555

Daniel-Escamilla committed:
docs(config): clarify STARSHIP_CONFIG supports one or multiple files

cappyzawa commented on src/context.rs:
```diff
@@ -462,20 +463,30 @@ impl Default for Context<'_> {
+            let default_path = home_dir(env)?.join(".config").join("starship.toml");
+            return Some(default_path.into_os_string());
+        }
+        // Devuelve siempre el valor de STARSHIP_CONFIG, sea uno o varios paths
```
Could you please update this comment to English for consistency with the rest of the codebase?

cappyzawa reviewed (commented):
Looking good overall, thanks!

Just a few minor suggestions to polish the details 🙏

Daniel-Escamilla committed:
docs: minor comment updates

Daniel-Escamilla commented:
Thank you for being so sharp! @cappyzawa

cappyzawa reviewed (dismissed):
LGTM 🥇 

Looking forward to using this feature when it's released. Great work! 👍

renovate[bot] committed:
build(deps): update embarkstudios/cargo-deny-action action to v2.0.13

Daniel-Escamilla committed:
Merge remote-tracking branch 'origin/master' into feature/config-flag-6555

Daniel-Escamilla committed:
Merge branch 'master' into feature/config-flag-6555

dzuelke commented:
I just realized that this breaks `starship config`, @Daniel-Escamilla and @cappyzawa:

```ShellSession
dzuelke@localhost starship % STARSHIP_CONFIG=docs/public/presets/toml/gruvbox-rainbow.toml:docs/public/presets/toml/nerd-font-symbols.toml target/debug/starship config     
Error: parsing as path with position docs/public/presets/toml/gruvbox-rainbow.toml:docs/public/presets/toml/nerd-font-symbols.toml

Caused by:
    No such file or directory (os error 2)
```

I guess the command should simply abort if `has_multiple_files` returns true?

dzuelke commented:
Also, may I suggest allowing an empty segment as a placeholder for the default config location (`~/.config/starship.toml`)? This is a common-ish pattern that other programs allow (e.g. `php` for `PHP_INI_SCAN_DIR`) to enable a "portable" way of injecting configs without having to know or specify the default config path.

That way, you could do this to prepend/append/surround other configs to the default config:

```ShellSession
dzuelke@localhost starship % # read prepend.toml before system default config
dzuelke@localhost starship % STARSHIP_CONFIG="prepend.toml:" starship …
```

```ShellSession
dzuelke@localhost starship % # read append.toml after system default config
dzuelke@localhost starship % STARSHIP_CONFIG=":docs/public/presets/toml/gruvbox-rainbow.toml" starship …
```

```ShellSession
dzuelke@localhost starship % # read defaults.toml before, then experiment.toml after, system default config
dzuelke@localhost starship % STARSHIP_CONFIG="defaults.toml::experiment.toml" starship …
```

cappyzawa commented:
@dzuelke Great catch! I apologize for missing this issue during my initial review.

Indeed, the current implementation breaks `starship config` when using multiple configuration files. The command incorrectly treats the entire path string (e.g., `"file1.toml:file2.toml"`) as a single filename, which could create unintended files.

## Proposed Solution

I think we should follow a pattern similar to kubectl's handling of `KUBECONFIG` with multiple files:

### For editing existing settings:
Track which file each setting originates from (LocationOfOrigin) and edit that file.

Example:
```toml
# file1.toml
[character]
success_symbol = "[>](green)"

# file2.toml  
[character]
error_symbol = "[x](red)"
```

```bash
STARSHIP_CONFIG="file1.toml:file2.toml"
starship config character.success_symbol "[✓](green)"  
# → Updates file1.toml (where success_symbol is defined)

starship config character.error_symbol "[✗](red)"
# → Updates file2.toml (where error_symbol is defined)
```

### For adding new settings:
Use a "default destination file" logic:
- First existing file in the list, OR
- Last path in the list (create if none exist)

Example:
```bash
# If file1.toml exists, file2.toml doesn't exist
STARSHIP_CONFIG="file1.toml:file2.toml:file3.toml"
starship config newmodule.enabled true
# → Adds to file1.toml (first existing file)

# If none exist
STARSHIP_CONFIG="file1.toml:file2.toml:file3.toml"
starship config newmodule.enabled true
# → Creates file3.toml and adds there (last path)
```

### Why this approach?
1. Preserves user intent - Settings stay in their original files
2. Predictable behavior - Clear rules for where changes go
3. Immediate effect - Changes are applied with correct precedence

### Implementation thoughts:

For this approach, we'd probably need to:

1. Track origins during merge: When `merge_toml_tables()` processes files, maintain a map of `config_key -> source_file_path` for each setting.

2. Update `get_configuration_edit()`: Instead of merging everything into a single document, determine which file to edit:
```rust
pub fn get_configuration_edit(context: &Context, key: &str) -> (DocumentMut, PathBuf) {
    let config_paths = parse_input_paths(context.get_config_path_os()?);
    
    // Check where this key is currently defined
    if let Some(origin_path) = find_key_origin(&config_paths, key) {
        // Edit the file where key exists
        let content = fs::read_to_string(&origin_path)?;
        (content.parse::<DocumentMut>()?, origin_path)
    } else {
        // New key: use default destination
        let dest_path = get_default_destination(&config_paths);
        let content = fs::read_to_string(&dest_path).unwrap_or_default();
        (content.parse::<DocumentMut>()?, dest_path)
    }
}
```

3. Modify `write_configuration()`: Accept a target file path parameter so it can write to the correct file instead of always using the merged path.

4. Helper function for default destination:
```rust
fn get_default_destination(paths: &[PathBuf]) -> PathBuf {
    // Find first existing file, or use last path if none exist
    paths.iter()
        .find(|p| p.exists())
        .unwrap_or(&paths[paths.len() - 1])
        .clone()
}
```

Let me know what you think about this approach, or if you have other ideas for implementation!

@Daniel-Escamilla What do you think about implementing this LocationOfOrigin-based approach? It's more complex than simply blocking multi-file configs, but provides a much better user experience.

cappyzawa commented:
@starship/astronauts 

I want to know if this LocationOfOrigin-based approach aligns with the project's direction. Would you prefer:

1. The kubectl-style LocationOfOrigin tracking (more complex but better UX)
2. Simply blocking multi-file configs for the `config` command  
3. Support empty segments as @dzuelke suggested (like PHP's `PHP_INI_SCAN_DIR`) - e.g., `STARSHIP_CONFIG="base.toml::custom.toml"` where `::` represents the default config location
4. Another approach?

Want to make sure we're all aligned before asking for the implementation work. 🙏

cappyzawa commented:
@Daniel-Escamilla @dzuelke 

I agree with @dzuelke's point about reconsidering whether `starship config` needs to support multiple files at all.

The core value of this PR - being able to render prompts from multiple config files - is already a huge win! That functionality alone enables all the use cases @dzuelke mentioned (presets + overrides, project-specific configs, etc).

For the `starship config` command, I think simply blocking it when multiple files are configured would be sufficient:
- Prevents the file creation bug
- Simple implementation (the fix I showed earlier)
- Users can still enjoy multi-file configs for prompt rendering
- They can edit individual files directly when needed

While the empty segment feature is interesting, I'd prefer to keep the current simple `:` separator behavior for now. Given that most users who use multiple configs probably know which file they want to edit anyway, the simple blocking approach seems reasonable.

What do you think about just going with the simple blocking approach for now?

Daniel-Escamilla commented:
@cappyzawa @dzuelke

I forgot that the main issue here is handling multiple files. Blocking starship config for multiple files might be the best option so this PR can be approved, and later we could open another issue to add support for multiple files in starship config.

Daniel-Escamilla committed:
feat(configure): block `starship config` when multiple files are configured

dzuelke commented:
Oh yeah totally, it just sounded like you wanted this solved before a merge. I agree that blocking `starship config` entirely for now on `if has_multiple_files()` is totally fine.

I think the empty segment functionality would still be useful, I think `parse_input_paths()` just needs another `map` to translate empty strings to the default config location.

However, maybe it's not a bad idea to also leave that out for now, because it'd also be useful if `$XDG_CONFIG_HOME/starship.toml` or `$XDG_CONFIG_HOME/starship/config.toml` was checked before `$HOME/.config`. That's when the empty segment starts getting really useful, because a project `README` could say "for Starship integration, do `export STARSHIP_CONFIG=:$(pwd)/conf/starship.toml` in the project root".

So, yeah, keep it simple for now, and in a future PR (happy to help implement, I'm comfy enough in Rust), add `XDG_CONFIG_HOME` lookup, empty `STARSHIP_CONFIG` segments, and `starship config` support for the empty segment.

cappyzawa reviewed (approved):
LGTM👍

cappyzawa commented:
@Daniel-Escamilla 
Sorry, I missed this in my review! I tested the current implementation and found that the fix is incomplete.

While `starship config` (editor mode) is now properly protected, `starship config name value` (direct value setting) still creates the unwanted files.

```bash
# Test reproduction:
echo '[character]\nsuccess_symbol = "❯"' > /tmp/test1.toml
echo '[directory]\ntruncation_length = 5' > /tmp/test2.toml

# This still fails
STARSHIP_CONFIG="/tmp/test1.toml:/tmp/test2.toml" starship config character.error_symbol "✗"
# Error: tries to create "/tmp/test1.toml:/tmp/test2.toml" file
```

The current fix only covers `edit_configuration()` but the same check needs to be added to `get_configuration_edit()` (line 233) and `write_configuration()` (line 243), which are used by `starship config name value` commands.

To avoid code duplication, consider creating a helper function:

```rust
fn ensure_single_config_file(context: &Context) -> Result<(), Box<dyn std::error::Error>> {
    if let Some(config_path) = context.get_config_path_os() {
        if StarshipConfig::has_multiple_files(&config_path.to_string_lossy()) {
            eprintln!("Error: starship config does not support editing multiple configuration files");
            eprintln!("Please edit the individual files directly");
            process::exit(1);
        }
    }
    Ok(())
}
```

Then call it from all three functions (`get_configuration_edit`, `write_configuration`, and `edit_configuration`).

Daniel-Escamilla committed:
feat(configure): block `starship config` and `starship config <NAME> <VALUE>` when multiple files are configured

Daniel-Escamilla committed:
refactor(config): centralize default config path and avoid duplication

cappyzawa reviewed (approved):
LGTM 🥇

Daniel-Escamilla committed:
refactor(bug_report): delegate to shared config helpers for file handling

Daniel-Escamilla committed:
refactor(config): remove unnecessary Option parameters and use shared facilities

Daniel-Escamilla commented on src/config.rs:
```diff
@@ -135,45 +135,132 @@ impl StarshipConfig {
+    /// Supports complex TOML structures including arrays, nested tables, and matrices
+    pub fn merge_config_files_runtime(config_paths: &str) -> Option<String> {
+        let paths = Self::parse_input_paths(config_paths);
+        let existing_paths: Vec<_> = paths.iter().filter(|p| p.exists()).collect();
+        if existing_paths.is_empty() {
+            log::warn!("No configuration files found for merging");
             return None;
         }
```
I see your point. The reason I kept the existence check is to avoid misleading warnings.
Without it, when files don’t exist the message becomes:
``` bash
[WARN] - No valid TOML content found in any configuration file
```
which looks like a parsing issue.

With the check, the log is:
``` bash
[WARN] - No configuration files found for merging
```
which better reflects the situation.

dzuelke replied:
But should that case produce a warning? I'd argue that
1. if `STARSHIP_CONFIG` is intentionally empty, then no warning should be produced;
2. if the default config location is used but the file doesn't exist, no warning should be produced (or maybe just an `INFO` level message?);
3. if a file specified in `STARSHIP_CONFIG` doesn't exist, it'd be useful to print the path in the warning message

davidkna replied:
To avoid this, this should perhaps just return a `toml::Table` or something like that.

Daniel-Escamilla commented on src/modules/custom.rs:
Ok. I'll handle the Windows-specific changes in a separate PR to keep this one focused on its original purpose.

davidkna replied:
Actually revert the changes please.

Daniel-Escamilla commented on src/bug_report.rs:
```diff
@@ -236,16 +234,51 @@ fn get_config_path(shell: &str) -> Option<PathBuf> {
 }
 
 fn get_starship_config() -> String {
```
Good point — I agree. I'll use the shared helper to obtain the combined configuration instead of duplicating the logic here. I'll make a small, incremental change to delegate to the central helper.

Daniel-Escamilla committed:
feat: unify config file handling and update tests for consistent behavior

Daniel-Escamilla committed:
refactor(bug_report): simplify config display and improve tests

Daniel-Escamilla committed:
refactor(context): remove comment to avoid merge problems

davidkna commented on src/context.rs:
```diff
@@ -463,19 +464,28 @@ impl Default for Context<'_> {
 fn get_config_path_os(env: &Env) -> Option<OsString> {
-    if let Some(config_path) = env.get_env_os("STARSHIP_CONFIG") {
-        return Some(config_path);
+    if let Some(config_line) = env.get_env_os("STARSHIP_CONFIG") {
+        let config_str = config_line.to_str()?;
+        if config_str.is_empty() {
+            let default_path = home_dir(env)?.join(".config").join("starship.toml");
+            return Some(default_path.into_os_string());
+        }
```
This avoids having two paths map to the default value, and removes the string conversion.

```suggestion
    if let Some(config_line) = env.get_env_os("STARSHIP_CONFIG")
        .filter(|v| !v.is_empty())
```

davidkna commented on src/config.rs:
```diff
@@ -134,46 +134,129 @@ impl StarshipConfig {
 
     /// Create a config from a starship configuration file
     fn config_from_file(config_file_path: Option<&OsStr>) -> Option<toml::Table> {
+        let config_file_path = config_file_path?;
```
```suggestion
   fn config_from_file(config_file_path: &OsStr) -> Option<toml::Table> {
```

davidkna commented on src/bug_report.rs:
```diff
@@ -308,4 +305,55 @@ mod tests {
     }
+    #[test]
+    fn test_get_starship_config_env_not_set() {
+        unsafe {
+            std::env::remove_var("STARSHIP_CONFIG");
+        }
```
This is `unsafe` for a reason. Please use `Env`/`Context` for env var mocks.

davidkna commented on src/config.rs:
```diff
@@ -134,46 +134,129 @@ impl StarshipConfig {
+    }
+
+    /// Parses input line to extract and expand file paths using std::env::split_paths
+    pub fn parse_input_paths(line: &str) -> Vec<std::path::PathBuf> {
```
```suggestion
    pub fn parse_input_paths(line: &OsStr) -> Vec<std::path::PathBuf> {
```

davidkna commented on src/config.rs:
```diff
@@ -134,46 +134,129 @@ impl StarshipConfig {
+        config_file_path: &OsStr,
+        context: Option<&Context>,
+    ) -> Option<String> {
+        Self::merge_config_files_runtime(config_file_path).or_else(|| {
+            let default_path = context
+                .and_then(|ctx| {
+                    ctx.get_home()
+                        .and_then(|home| utils::default_starship_config_path(Some(home.as_path())))
+                })
+                .or_else(|| utils::default_starship_config_path(None))?;
+            utils::read_file(&default_path).ok()
```
Shouldn't `get_config_path_os` already handle fallbacks?

Daniel-Escamilla committed:
Revert "fix: add UTF-8 codepage setting for Windows CMD commands in custom modules"

This reverts commit fc1e3dcae55f11745348ed534fc3b3a0c46aaade.

Daniel-Escamilla commented on src/context.rs:
```diff
@@ -462,20 +463,30 @@ impl Default for Context<'_> {
-        if let Some(home) = env.get_env("HOME") {
-            return Some(PathBuf::from(home));
-        }
+    // Try to get HOME from environment first (mocked or real)
+    if let Some(home) = env.get_env("HOME") {
+        return Some(PathBuf::from(home));
     }
```
Apologies, I wasn’t aware of this behavior. I’ll revert the change.

Daniel-Escamilla committed:
Revert home_dir change

Daniel-Escamilla committed:
Remove unsafe env access; use Env/Context in tests for deterministic behavior

Daniel-Escamilla committed:
Refactor config API: accept &OsStr, add is_explicit_config flag, and improve merge behavior

Daniel-Escamilla committed:
refactor(config): accept &OsStr in config_from_file; update parse_input_paths and callers

Daniel-Escamilla commented:
Hi @davidkna. Just wanted to kindly check in on the status of this PR. It’s been open for a while, and I wanted to know if there’s anything else I should update or adjust to help move it forward. Thanks

dzuelke commented:
Because you’re opposed to the idea, or just the implementation, or want someone else to review, @davidkna ?

Daniel-Escamilla committed:
Merge branch 'master' into feature/config-flag-6555
